### DISCUSSION TITLE
RP2040/Pico USB serial host

### DISCUSSION THREAD

**Discussion by Hexxeh** [hathach/tinyusb#925] [Q&A]
I'm trying to have an RP2040/Pico act as a USB host for a USB serial adaptor. I can get the adaptor to be detected, but attempting to use `tuh_cdc_receive` results in a `stalled` error code, and `tuh_cdc_connect` returns `false` because the `is mounted` check fails.

```
USBH init
CDC init
USBH DEVICE ATTACH
Open EP Control with Size = 8
Get 8 byte of Device Descriptor
Control Setup: 80 06 00 01 00 00 08 00 
on EP 00 with 8 bytes
on EP 80 with 8 bytes
Control data:
  0000:  12 01 10 01 00 00 00 40                          |.......@|
on EP 00 with 0 bytes

Port reset 
Set Address = 1
Control Setup: 00 05 01 00 00 00 00 00 
on EP 00 with 8 bytes
on EP 80 with 0 bytes

Open EP Control with Size = 64
Get Device Descriptor
Control Setup: 80 06 00 01 00 00 12 00 
on EP 00 with 8 bytes
on EP 80 with 18 bytes
Control data:
  0000:  12 01 10 01 00 00 00 40 C4 10 60 EA 00 01 01 02  |.......@..`.....|
  0010:  03 01                                            |..|
on EP 00 with 0 bytes

Get 9 bytes of Configuration Descriptor
Control Setup: 80 06 00 02 00 00 09 00 
on EP 00 with 8 bytes
on EP 80 with 9 bytes
Control data:
  0000:  09 02 20 00 01 01 00 80 32                       |.. .....2|
on EP 00 with 0 bytes

Get Configuration Descriptor
Control Setup: 80 06 00 02 00 00 20 00 
on EP 00 with 8 bytes
on EP 80 with 32 bytes
Control data:
  0000:  09 02 20 00 01 01 00 80 32 09 04 00 00 02 FF 00  |.. .....2.......|
  0010:  00 02 07 05 81 02 40 00 00 07 05 01 02 40 00 00  |......@......@..|
on EP 00 with 0 bytes

Set Configuration = 1
Control Setup: 00 09 01 00 00 00 00 00 
on EP 00 with 8 bytes
on EP 80 with 0 bytes

Device configured
A device with address 1 is mounted
Control Setup: 21 22 03 00 00 00 00 00 
connect: 1
on EP 00 with 8 bytes
on EP 80 with 0 bytes
Control failed: result = 2

connected cdc: 2
```

Is what I'm trying to do here supported, and if not, what's missing that I could try to add support for?

**Comment by chegewara** [comment]
From my experience, not related to tinyusb, some CDC devices does not accept when you set device configuration 1, but accept configuration 0 (which is wrong). Another thing that some devices require is to send set control line (DTR/RTS) and set line codding (probably optional).

**Reply by Hexxeh** [reply]
Where does the configuration get set? I'm not using either explicitly. Pretty bare bones code:

```
extern void cdc_task(void);

void core1_entry() {
    board_init();
    tusb_init();
    
    while (1) {
        tuh_task();
        cdc_task();
    }
}

CFG_TUSB_MEM_SECTION static char serial_in_buffer[64] = { 0 };

bool handle_connected(uint8_t dev_addr, tusb_control_request_t const * request, xfer_result_t result) {
    printf("connected cdc: %d\n", result);
}

void tuh_mount_cb(uint8_t dev_addr)
{
  printf("A device with address %d is mounted\n", dev_addr);
  tuh_cdc_receive(dev_addr, serial_in_buffer, sizeof(serial_in_buffer), true); // waiting for next data
}

void tuh_umount_cb(uint8_t dev_addr)
{
  printf("A device with address %d is unmounted \r\n", dev_addr);
}

void tuh_cdc_xfer_isr(uint8_t dev_addr, xfer_result_t event, cdc_pipeid_t pipe_id, uint32_t xferred_bytes)
{
  (void) event;
  (void) pipe_id;
  (void) xferred_bytes;

  printf("CDC recv: %d\n", xferred_bytes);
  printf(serial_in_buffer);
  tu_memclr(serial_in_buffer, sizeof(serial_in_buffer));

  tuh_cdc_receive(dev_addr, serial_in_buffer, sizeof(serial_in_buffer), true); // waiting for next data
}

void cdc_task(void)
{
}
```

**Reply by Hexxeh** [reply]
Swapped out the USB serial adaptor for another Pico, and that gets a bit further. I see extra endpoints being listed.

Made a slight patch:

```
  // all interface are configured
  if (itf_num == sizeof(dev->itf2drv) || true)
```

and now I can receive data over USB from a second Pico. Seems to be the other device isn't enumerating properly and is missing endpoints perhaps? It's a CP2102 clone.

**Reply by hathach** [reply]
Cp201x and fdti is not cdc, those are proprietary serial. 

Note: host cdc API may got rework in the future

**Reply by Hexxeh** [reply]
Ah, that would be why! Apologies, my understanding of the USB stack isn't so good.

Any thoughts on the patch I had to make? There's another one for the `pipe_id` in `tuh_cdc_xfer_isr` always being the same value, but I can submit a PR for that.

**Reply by hathach** [reply]
@Hexxeh that is ok, don't worry about the PR. The CDC really needs a rework with probably lots of API changes. I have manged to do so with msc and hid. Will do that with CDC later on. 

Note: For host rp2040, there could be an issue using bulk (cdc), since its controller can only carry only ONE bulk or control at any one time. Which is not what the current host stack would expect. Anyway, please expect a significant changes in host cdc API.

**Reply by Hexxeh** [reply]
I made good progress reading from device to host (after an initial host to device message), but then today when I added a second host to device message I hit a panic:

```
CDC xfer callback
  Queue EP 82 with 64 bytes ... OK
on EP 82 with 1 bytes
CDC xfer callback
  Queue EP 82 with 64 bytes ... OK
on EP 82 with 1 bytes
CDC xfer callback
  Queue EP 82 with 64 bytes ... OK
  Queue EP 02 with 10 bytes ...
*** PANIC ***

ep 2 out was already available
```

Is this the bulk/control concurrency issue you mention? If so, is there a workaround?

My process looks like this:

- Host->Device: init message
- Device->Host: data message (repeated every 100ms)
- Host->Device: ack message (at least every 2000ms)

I hit the panic when sending the ack message, either from `cdc_task` in the main event loop or a timer callback.

This is using a commit (5a4fc1151a7e62a291a90836165942737a1a2d6f) prior to the double buffering changes, with the latest master as of today it fails much earlier:

```
USBH init          
CDC init
hcd_init 0
ili9341: init done
USBH DEVICE ATTACH
hcd_port_connect_status
Open EP Control with Size = 8
hcd_edpt_open dev_addr 0, ep_addr 0
hw_endpoint_init dev 0 ep 0 out xfer 0
dev 0 ep 0 out setup buffer @ 0x50100180
endpoint control (0x50100100) <- 0xa0000180
Get 8 byte of Device Descriptor
Control Setup (addr = 0): 80 06 00 01 00 00 08 00 
hw_endpoint_init dev 0 ep 0 out xfer 0
dev 0 ep 0 out setup buffer @ 0x50100180
endpoint control (0x50100100) <- 0xa0000180
Transfer complete
Sent setup packet
on EP 00 with 0 bytes
hcd_edpt_xfer dev_addr 0, ep_addr 0x80, len 8
hw_endpoint_init dev 0 ep 0 in xfer 0
dev 0 ep 0 in setup buffer @ 0x50100180
endpoint control (0x50100100) <- 0xa0000180
Prepare Buffer Control:
  Buffer Control 1 0x0: len = 0, available = 0, full = 0, last = 0, stall = 0, reset = 0, toggle = 0
  Buffer Control 0 0x7408: len = 8, available = 1, full = 0, last = 1, stall = 0, reset = 1, toggle = 1
Buffer complete
buf_status 0x00000001
Single Buffered: ep_ctrl = A0000180
_hw_endpoint_xfer_sync:
  Buffer Control 1 0x0: len = 0, available = 0, full = 0, last = 0, stall = 0, reset = 0, toggle = 0
  Buffer Control 0 0xe008: len = 8, available = 0, full = 1, last = 1, stall = 0, reset = 0, toggle = 1
Completed transfer of 8 bytes on ep 0 in
Transfer complete
assertion "ep->active" failed: file "/home/hexxeh/pico/pico-sdk/lib/tinyusb/src/portable/raspberrypi/rp2040/hcd_rp2040.c", line 165, function: hw_trans_complete
```

**Reply by Hexxeh** [reply]
@hathach would a CDC ping/pong demo app for two Picos help in fixing this, or is it clearer to you what's wrong anyway? Happy to put one together if it'd help.

**Reply by hathach** [reply]
@Hexxeh the above log show that device doesn't get pass the very first command of enumeration (Get 8 byte of device descriptor) just yet. It shouldn't failed at this stage. Is the device you connecting to another pico running device stack, if so, could you upload device firmware or uf2 here so that I could try to reproduce the issue.

**Reply by Hexxeh** [reply]
Yup it's a Pico running the device stack. I'll upload the U2F/source. The log further up is the same code, just with an older version of TinyUSB on the host side.

**Reply by Hexxeh** [reply]
[simulator.zip](https://github.com/hathach/tinyusb/files/6783403/simulator.zip)

For some reason GitHub doesn't accept the U2F extension, but does accept ZIP, so it's in there. Here's the matching source (drops straight into an empty Pico SDK project): https://gist.github.com/Hexxeh/8e5fe732bfc424609828582537d8261a

**Comment by johnkjellberg** [comment]
@Hexxeh How far did you get with this? Where you able to resolve it? Do you have some source code? I really need to get it working. So any help would be appreciated. I got the mounting working with your fix `if (itf_num == sizeof(dev->itf2drv) || true)`.

But `tuh_cdc_send` failes with an error.  To me it seems like it overwrites crucial parts of the enumeration due to this check ` if (transfer_type == TUSB_XFER_INTERRUPT)` in `hcd_rp2040.c`. Changing it to `if (transfer_type != TUSB_XFER_CONTROL)` fixes that part. But doesn't seem like it is actually sending any data when I do `tuh_cdc_send`.
I'm quite new to USB, so I'm not really sure how it is supposed to work.

**Reply by Hexxeh** [reply]
I didn't get any further unfortunately. Ended up using a Zero instead and eating the slower boot time to get the project done. I'd like to swap back though should this get fixed.

**Reply by johnkjellberg** [reply]
> I didn't get any further unfortunately. Ended up using a Zero instead and eating the slower boot time to get the project done. I'd like to swap back though should this get fixed.

Ok!

<s>It is actually a bit strange. As far as I can tell it really can't work. Class CDC (0x02) is for control, and CDC Data(0x0A) is for sending the data as I understand it. But during enumeration, 0x0A is defined as not supported. So I'm confused that functions for sending and receiving data is implemented without setting up of the required device is.

So it might be that just enumerating the CDC-Data (0x0A) interface is needed, and then the `tuh_cdc_send` can be used on that interface.</s>
Edit: I missed that this was actually handled in the code. But the error message was a bit missleading.

**Reply by johnkjellberg** [reply]
Actually. My best guess now is that since CDC is actually two parts, CDC and CDC Data and the RP2040 layer can only store one instance of a EP that isn't interrupt transfer type, the bulk EP gets overwritten during the enumeration phase. 

I guess that is what @Hexxeh meant with `For host rp2040, there could be an issue using bulk (cdc), since its controller can only carry only ONE bulk or control at any one time`. I missunderstood that.

I will have to see how difficult it would be to make sure things aren't overwritten during the enumeration. In theory, it shouldn't be too difficult.

**Reply by johnkjellberg** [reply]
I got this working now! What needs to be changed is to support `TUSB_XFER_BULK` by using the EPx register for sending/receiving data, but also make sure to store the endpoints information regarding those somewhere. Right now I'm using the `ep_pool` with `struct hw_endpoint` for that. But is is quite wasteful, so I'll probably create a new struct and pool specifically for bulk transfers unless it requires to much special handling at different places, I'll see.

**Reply by johnkjellberg** [reply]
Actually. Now I found this pull request https://github.com/hathach/tinyusb/pull/1434 that fixes the issue and works for CDC host!

The only thing not working is the detection of mounting for CDC. I.e. `if (itf_num == CFG_TUH_INTERFACE_MAX)` in file `usbh.c`.

**Reply by GraDKh** [reply]
@johnkjellberg 
Did you get mount callback working for CDC?

**Reply by johnkjellberg** [reply]
> @johnkjellberg Did you get mount callback working for CDC?

No, I didn't really bother with that. My application could do a workaround by manually checking `tuh_mounted` instead.
